From 7bb968c57f828577a81871179e61d7c8ce2854c6 Mon Sep 17 00:00:00 2001
From: Amit Shah <amit.shah@redhat.com>
Date: Mon, 18 Aug 2014 10:00:28 -0400
Subject: [hwrng] Pass entropy to add_hwgenerator_randomness() in bits, not bytes

Message-id: <0b99a500b46ad95b0901107b0c6cba8b89a072cf.1408347166.git.amit.shah@redhat.com>
Patchwork-id: 87842
O-Subject: [RHEL7.1 kernel PATCH 5/6] hwrng: Pass entropy to add_hwgenerator_randomness() in bits, not bytes
Bugzilla: 1129195
RH-Acked-by: Amos Kong <akong@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Herbert Xu <herbert.xu@redhat.com>

From: Stephen Boyd <sboyd@codeaurora.org>

rng_get_data() returns the number of bytes read from the hardware.
The entropy argument to add_hwgenerator_randomness() is passed
directly to credit_entropy_bits() so we should be passing the
number of bits, not bytes here.

Fixes: be4000bc464 "hwrng: create filler thread"
Acked-by: Torsten Duwe <duwe@suse.de>
Signed-off-by: Stephen Boyd <sboyd@codeaurora.org>
Signed-off-by: Theodore Ts'o <tytso@mit.edu>
(cherry picked from commit e02b876597777ab26288dd2611a97b597d14d661)
Signed-off-by: Amit Shah <amit.shah@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/char/hw_random/core.c b/drivers/char/hw_random/core.c
index f9ae152..52a6c00 100644
--- a/drivers/char/hw_random/core.c
+++ b/drivers/char/hw_random/core.c
@@ -357,7 +357,7 @@ static int hwrng_fillfn(void *unused)
    continue;
   }
   add_hwgenerator_randomness((void *)rng_fillbuf, rc,
-        (rc*current_quality)>>10);
+        rc * current_quality * 8 >> 10);
  }
  hwrng_fill = NULL;
  return 0;
-- 
1.7.1